### DISCUSSION TITLE
Use Conveyor in a Compose Multiplatform module targeting Android, iOS and JVM

### DISCUSSION THREAD

**Discussion by vinceglb** [hydraulic-software/conveyor#97] [Q&A] [answered]
I'm trying to setup Conveyor in my project. 

When I run `./gradlew printConveyorConfig`, it throws this error:

```
> Task :composeApp:printConveyorConfig FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':composeApp:printConveyorConfig'.
> Task with name 'jar' not found in project ':composeApp'.
```

I think I'm getting the error because I added Conveyor to a module targeting Android, iOS and JVM and the `jar` command does not exist in this config. Here is my config:

```kotlin
plugins {
    alias(libs.plugins.kotlinMultiplatform)
    alias(libs.plugins.androidApplication)
    alias(libs.plugins.jetbrainsCompose)
    alias(libs.plugins.conveyor)
}

group = "..."
version = "0.1.0"

kotlin {
    androidTarget {
        compilations.all {
            kotlinOptions {
                jvmTarget = "17"
            }
        }
    }

    jvm("desktop")
    jvmToolchain(17)

    listOf(
        iosX64(),
        iosArm64(),
        iosSimulatorArm64()
    ).forEach { iosTarget ->
        iosTarget.binaries.framework {
            baseName = "ComposeApp"
            isStatic = true
        }
    }

   ...
}

android {
   ...
}

compose.desktop {
    application {
        mainClass = "MainKt"
   }
}

dependencies {
    linuxAmd64(compose.desktop.linux_x64)
    macAmd64(compose.desktop.macos_x64)
    macAarch64(compose.desktop.macos_arm64)
    windowsAmd64(compose.desktop.windows_x64)
}

// region Work around temporary Compose bugs.
configurations.all {
    attributes {
        attribute(Attribute.of("ui", String::class.java), "awt")
    }
}
// endregion
```

Is it possible to make it work in this configuration? 
Can I customize Conveyor to not use `./gradlew jar` but another command instead?
Do I have to create a specific module with JVM target only to make it work?

Thank you for your help 🙏

**Comment by mikehearn** [comment]
Hi,

Do you have a `jvmJar` target? The code for the Gradle plugin is open source and can be seen [here](https://github.com/hydraulic-software/conveyor/blob/master/gradle-plugin/src/main/kotlin/hydraulic/conveyor/gradle/ConveyorConfigTask.kt#L116). It looks first for `jvmJar` because that's what CfD uses, and only then for `jar`. Sounds like somehow you don't have either; perhaps Jetbrains changed something again.

At the moment the only way to change where the plugin looks would be to improve the code. However, you don't have to use the Gradle plugin at all. If you run `conveyor generate compose example-app` then run `./gradlew printConveyorConfig` inside, you'll see what's being extracted via the plugin is just ordinary configuration: some metadata, and the paths for the JARs for each platform. If we can't make it work with the Gradle plugin, then you always have the option to compute the config in some other way e.g. using custom Gradle code. It's not that much.

But we test CfD projects so I'd really like to know why it doesn't work for you. What version of CfD are you using?

**Reply by vinceglb** [reply]
Hi, thank you for your help!

I have the `jvmJar` target. When I run `./gradlew tasks`, I find this task in the list. And I can run it successfully. It looks like `project.tasks.findByName("jvmJar")` is returning null, but the command exists in my project.

Actually, I'm using:

```kotlin
agp = "8.2.2"
conveyor = "1.8"
compose-plugin = "1.6.0-beta01"
kotlin = "1.9.22"
```

**Reply by mikehearn** [reply]
Hmm, our template project uses compose 1.3.1 it seems. Perhaps something has changed in Gradle that means this task can no longer be looked up? Can you get a non-null result from `project.tasks.findByName("jvmJar")` in your own build.gradle file?

**Reply by mikehearn** [reply, marked as answer]
OK, the difference seems to be:

```
kotlin {
    jvm {
        withJava()
    }
```

which works, vs what you have:

```
kotlin {
    jvm("desktop")
}
```

which doesn't. I have no idea what the difference is meant to be, though. If you can figure out which task is the equivalent of `jvmJar` (basically, compile and package the app up into jars) then we can improve the plugin to understand yet another variant of CfD configuration.

**Reply by vinceglb** [reply]
Exactly! When passing an argument to `jvm()`, like `jvm("desktop")`, it renames the task. So `jvmJar` task is renamed to `desktopJar` in my project.

When using the default template of Compose Multiplatform, jvm target is renamed to "desktop". You can try here: https://kmp.jetbrains.com/

By replacing `jvm("desktop")` to `jvm()` in my project, it works like expected ✅

**Reply by mikehearn** [reply]
Ah. I see. I wonder if there's a better way to get the right name.

**Reply by vinceglb** [reply]
Yes, it would be perfect if `printConveyorConfig` could adapt to the jvm target name or at least have a parameter to customize the command it will call.

**Comment by hlinero** [comment]
Thank you @vinceglb and @mikehearn for the information here.

When I follow the steps from this discussion, that is

Change:
```
kotlin {
    jvm("desktop")
}
```

to :
```
kotlin {
     jvm {  withJava()  }
}
```

The error that I get when syncing or running `./gradlew jar` is:
```
> 'withJava()' is not compatible with Android Plugins
  Incompatible Android Plugin applied: 'com.android.application'

kotlin {
      jvm {
          withJava() /* <- cannot be used when the Android Plugin is present */
      }
  }

```

My project also targets `ios`, `Android`, and `Desktop`.

@mikehearn I do see the task `jvmJar` tasks when running `./gradlew tasks`.

Other info:
```
agp = "8.2.2"
compose = "1.6.6"
compose-plugin = "1.6.0"
kotlin = "1.9.22"
conveyor = "1.8"
```

Any suggestions on how to address this issue?

Thank you

**Reply by vinceglb** [reply]
@hlinero You can just remove `withJava()` and it will work.

```kotlin
kotlin {
     jvm()
}
```

Here a project targetting Android, iOS and JVM where I implemented Conveyor:
https://github.com/vinceglb/SpaceDrop/blob/74abde9657e8f505493331d5ebb3a7acf9992304/composeApp/build.gradle.kts#L21

**Reply by hlinero** [reply]
Thank you @vinceglb 

That error goes away but now I get a new one when running `./gradlew jar`

```
Task 'jar' not found in root project 'ProjectManagement' and its subprojects.
```

I have added the `alias(libs.plugins.conveyor)` in the `plugin` scope for the `build.gradle.kts` file of the `composeApp` module and  project level `build.gradle.kts` file too.

Looking at your project you also placed it at those two places. Am I missing something? Any suggestions?

**Reply by vinceglb** [reply]
Does this command work? `./gradlew jvmJar`

**Reply by hlinero** [reply]
Yes, it does. @vinceglb.

Now everything is working fine.

Thanks for the help.

**Reply by mikehearn** [reply]
Thanks @vinceglb !

**Reply by michaellee123** [reply]
after I deleted the "desktop" and then I  got an error message "java.lang.ClassNotFoundException: MainKt
", can you help me? thanks!
